packages/@o3r/extractors/builders/aggregate-migration-scripts/index.ts

```diff
@@ -0,0 +1,169 @@
+  !!libVersion.libVersion && !!libVersion.appVersion;
+
+/**
+ * Get the history of versions for each library in chronological order
+ * @param migrationFiles
+ */
+const getLibrariesVersions = (migrationFiles: MigrationFileEntry[]): LibraryVersions => {
+  const libraries: LibraryVersions = {};
+  for (const file of migrationFiles) {
+    if (file.migrationObject.libraries) {
+      for (const [lib, version] of Object.entries<string>(file.migrationObject.libraries)) {
+        if (!libraries[lib]) {
+          libraries[lib] = [];
+        }
+        const libVersion = {
+          libVersion: semver.coerce(version)?.raw,
+          appVersion: semver.coerce(file.migrationObject.version)?.raw
+        };
+
+        if (isValidLibVersion(libVersion)) {
+          libraries[lib].push(libVersion);
+        }
+      }
+    }
+  }
+  // Sort and dedupe versions
+  Object.entries(libraries).forEach(([libName, libVersions]) => {
+    libVersions.sort((a, b) => semver.compare(a.appVersion, b.appVersion));
+    libraries[libName] = libVersions.filter((libVersion, index) =>
+      index < 1 || libVersions[index - 1].libVersion !== libVersion.libVersion);
+  });
+  return libraries;
+};
+
+/**
+ * Write the migration files on disk
+ * @param migrationFiles
+ * @param destinationPath
+ */
+const writeMigrationFiles = async (migrationFiles: MigrationFileEntry[], destinationPath: string) => {
+  const outputDirectory = resolve(destinationPath);
+  if (fs.existsSync(outputDirectory)) {
+    await rm(outputDirectory, {recursive: true});
+  }
+
+  for (const file of migrationFiles) {
+    const distFilePath = join(outputDirectory, basename(file.fileName));
+    if (!fs.existsSync(dirname(distFilePath))) {
+      await mkdir(dirname(distFilePath), {recursive: true});
+    }
+    await writeFile(distFilePath, JSON.stringify(file.migrationObject, null, 2) + '\n');
+  }
+};
+
+/**
+ * Aggregate the relevant changes from libraries into the original migration files
+ * @param migrationFiles
+ * @param librariesVersions
+ * @param resolver
+ */
+const aggregateLibrariesChangesIntoMigrationFiles = async (migrationFiles: MigrationFileEntry[], librariesVersions: LibraryVersions, resolver: (lib: string) => string) => {
+  for (const file of migrationFiles) {
+    if (file.migrationObject.libraries) {
+      const appVersion = semver.coerce(file.migrationObject.version)?.raw;
+      for (const [lib, version] of Object.entries<string>(file.migrationObject.libraries)) {
+        const libIndex = librariesVersions[lib].findIndex((l) => l.appVersion === appVersion);
+        if (libIndex > 0) {
+          const newLibVersion = semver.coerce(version)?.raw;
+          const previousLibVersion = librariesVersions[lib][libIndex - 1].libVersion;
+          if (previousLibVersion !== newLibVersion) {
+            const libRange = new semver.Range(`>${previousLibVersion} <=${newLibVersion}`);
```

Comment:
Geniune question here: I am not sure I understand what we are doing:
`librariesVersion` contains the sorted list of all the version requested for each migration file.
For each migration file, we look for the previous version and look for the migration scripts in between, so why don't we just look for the migration script between the first requested version in `librariesVersion` and the latest one instead of this iteration ?

Reply:
From what I see we are sorting on `appVersion` and here we are checking the `libVersion`